src/grapheme.rs

```diff
@@ -85,351 +83,469 @@ impl<'a> Graphemes<'a> {
-                Start | Prepend => match cat {
-                    gr::GC_Control => {         // rule GB5
-                        take_curr = state == Start;
-                        break;
-                    }
-                    gr::GC_L => HangulL,
-                    gr::GC_LV | gr::GC_V => HangulLV,
-                    gr::GC_LVT | gr::GC_T => HangulLVT,
-                    gr::GC_Prepend if self.extended => Prepend,
-                    gr::GC_Regional_Indicator => Regional,
-                    gr::GC_E_Base | gr::GC_E_Base_GAZ => Emoji,
-                    _ => FindExtend
-                },
-                FindExtend => {         // found non-extending when looking for extending
-                    take_curr = false;
-                    break;
-                },
-                HangulL => match cat {      // rule GB6: L x (L|V|LV|LVT)
-                    gr::GC_L => continue,
-                    gr::GC_LV | gr::GC_V => HangulLV,
-                    gr::GC_LVT => HangulLVT,
-                    _ => {
-                        take_curr = false;
-                        break;
-                    }
-                },
-                HangulLV => match cat {     // rule GB7: (LV|V) x (V|T)
-                    gr::GC_V => continue,
-                    gr::GC_T => HangulLVT,
-                    _ => {
-                        take_curr = false;
-                        break;
-                    }
-                },
-                HangulLVT => match cat {    // rule GB8: (LVT|T) x T
-                    gr::GC_T => continue,
-                    _ => {
-                        take_curr = false;
-                        break;
-                    }
-                },
-                Regional => match cat {     // rule GB12/GB13
-                    gr::GC_Regional_Indicator => FindExtend,
-                    _ => {
-                        take_curr = false;
-                        break;
-                    }
-                },
-                Emoji => match cat {        // rule GB10: (E_Base|EBG) Extend* x E_Modifier
-                    gr::GC_E_Modifier => continue,
-                    _ => {
-                        take_curr = false;
-                        break;
-                    }
-                },
-                Zwj => match cat {          // rule GB11: ZWJ x (GAZ|EBG)
-                    gr::GC_Glue_After_Zwj => continue,
-                    gr::GC_E_Base_GAZ => Emoji,
-                    _ => {
-                        take_curr = false;
-                        break;
-                    }
-                },
-            }
-        }
+#[derive(PartialEq, Eq)]
+enum PairResult {
+    NotBreak,  // definitely not a break
+    Break,  // definitely a break
+    Extended,  // a break if not in extended mode
+    CheckCrlf,  // a break unless it's a CR LF pair
+    Regional,  // a break if preceded by an even number of RIS
+    Emoji,  // a break if preceded by emoji base and extend
```

Comment:
I changed this to read "a break iff not in extended mode" which I think is more precise. What you suggest is the converse of what I originally said, but both are true.